eth/rlp/chain_gaps.py

```diff
@@ -0,0 +1,40 @@
+from typing import Tuple, Iterable
+
+import rlp.sedes
+from eth_utils import to_tuple
+
+from eth._utils.numeric import signed_to_unsigned, unsigned_to_signed
+from eth.rlp.sedes import uint32
+from eth.typing import BlockRange
+
+chain_gaps = rlp.sedes.CountableList(rlp.sedes.List((uint32, uint32)))
+
+# Chain gaps are defined as sequence of markers that define gaps in a chain of connected
+# entities. The right hand side of the very last marker is expected to be -1, meaning the gap
+# is open-ended. E.g. (500, -1) means: Every header from number 500 upwards is missing.
+# [[first_missing, last_missing], ..., [first_missing, -1]]
+# Since RLP doesn't define signed integers, we convert the right-hand side from signed_to_unsigned
+# before entries are written and convert from unsigned_to_signed after entries are read from disk.
```

Comment:
Oh, I think your proposal is great! It's a really nice improvement to tighten the representation of the chain integrity.

Two examples to double check I got it right:

Now -> Then

1. An empty chain with only genesis

`[[1, -1] ]` -> `[[], 1]`

2. A chain with Checkpoint header 250 and 500 and no other headers

`[[1, 249], [251, 499], [501, -1]]` -> `[[1, 249], [251, 499], 501]`

>and I figure you might like that the types can be tightened from `int` to `BlockNumber`

I didn't do that to avoid bloating the calculations with things like `new_tail = ((highest_missing_number, gap_end), (newly_persisted + BlockNumber(1), BlockNumber(-1)),)` but since we got rid of the `-1` part, this may now become practical :+1:

Reply:
Ah yeah, I just noticed your example further down `(GapChange.GapSplit, (((2, 2), (4, 4)), 6))` which indicates we are on the same page :+1: